File: RDD/Service/DataCollectorServiceRDD.vb
Imports IDAUtil

Public Class DataCollectorServiceRDD
    Private ReadOnly dataCollectorServer As DataCollectorServer
    Private ReadOnly dataCollectorSap As DataCollectorSap

    Public Sub New(dataCollectorServer As DataCollectorServer, dataCollectorSap As DataCollectorSap)
        Me.dataCollectorServer = dataCollectorServer
        Me.dataCollectorSap = dataCollectorSap
    End Sub

    Public Function GetRddList(salesOrg As String, id As String) As List(Of RddOutputBean)

        Dim ZVList As List(Of ZV04HNProperty)
        ZVList = dataCollectorSap.getZV04HNList(salesOrg, IDAEnum.Task.rdd)

        If ZVList Is Nothing Then
            Return Nothing
        Else
            ZVList = ZVList.Where(Function(x) getZVCondition(x, salesOrg)).ToList
        End If

        Dim customerDataList As List(Of CustomerDataProperty) = dataCollectorServer.getCustomerDataList(salesOrg)
        Dim query As List(Of RddOutputBean)

        query = (
                From zv In ZVList
                Join cm In customerDataList On zv.shipto Equals cm.shipTo And zv.soldto Equals cm.soldTo
                Select New RddOutputBean(zv, cm, id)
                ).ToList()

        Return query
    End Function

    Public Function getBHList(salesOrg As String) As List(Of BankHolidayProperty)
        Return dataCollectorServer.getBHList(salesOrg)
    End Function

    Private Function getZVCondition(zv As ZV04HNProperty, salesOrg As String) As Boolean
        Dim flag As Boolean = zv.delBlock <> IDAConsts.bypassOrderChangesBlock

        Select Case salesOrg
            Case "ZA01", "NG01", "KE02"
                flag = flag AndAlso zv.delBlock <> "Z4" AndAlso zv.delBlock <> "04"
            Case "ES01", "PT01"
                flag = flag AndAlso zv.delBlock <> "ZG"
            Case Else
        End Select

        Return flag
    End Function
End Class

File: RDD/Service/DataCollectorServiceRDD.cs
using IDAUtil;
using IDAUtil.Model.Properties.ServerProperty;
using IDAUtil.Model.Properties.TcodeProperty.ZV04Obj;
using IDAUtil.Service;
using System.Collections.Generic;
using System.Data;
using System.Linq;

namespace RDD {
    public class DataCollectorServiceRDD {
        private readonly DataCollectorServer dataCollectorServer;
        private readonly DataCollectorSap dataCollectorSap;

        public DataCollectorServiceRDD(DataCollectorServer dataCollectorServer, DataCollectorSap dataCollectorSap) {
            this.dataCollectorServer = dataCollectorServer;
            this.dataCollectorSap = dataCollectorSap;
        }

        public List<RddOutputBean> GetRddList(string salesOrg, string id) {

            var ZVList = dataCollectorSap.getZV04HNList(salesOrg, IDAEnum.Task.rdd);

            if (ZVList is null) {
                return null;
            } else {
                ZVList = ZVList.Where(x => getZVCondition(x, salesOrg)).ToList();
            }

            var customerDataList = dataCollectorServer.getCustomerDataList(salesOrg);

            var query = (from zv in ZVList
                          join cm in customerDataList on new { key0 = zv.shipto, key1 = zv.soldto } equals new { key0 = cm.shipTo, key1 = cm.soldTo }
                          select new RddOutputBean(zv, cm, id)).ToList();
            return query;
        }

        public List<BankHolidayProperty> getBHList(string salesOrg) {
            return dataCollectorServer.getBHList(salesOrg);
        }

        private bool getZVCondition(ZV04HNProperty zv, string salesOrg) {

            bool flag = (zv.delBlock ?? "") != IDAConsts.DelBlocks.leadTimeBlock;

            switch (salesOrg) {
                case "ZA01":
                case "NG01":
                case "KE02": {
                        flag = flag && (zv.delBlock ?? "") != "Z4" && (zv.delBlock ?? "") != "04";
                        break;
                    }

                case "ES01":
                case "PT01": {
                        flag = flag && (zv.delBlock ?? "") != "ZG";
                        break;
                    }

                case "PL01":
                case "CZ01": {
                        flag = flag && (zv.delBlock ?? "") != "Z8" && (zv.delBlock ?? "") != "ZV" && (zv.delBlock ?? "") != "ZW";
                        break;
                    }
                case "RO01": {
                        flag = flag && zv.route != "ROEMGY";
                        break;
                    }

                default: {
                        break;
                    }
            }

            return flag;
        }
    }
}